### ISSUE TITLE
The issue of model evaluation results

### ISSUE THREAD

**Issue by tang12138-yyhh** [eml-eda/mixprec-pruning#3]
1. Hello, I would like to ask about the evaluation of the results after the quantization and pruning (icl_mixprec_training_cost_size.py) has been completed, which has generated .pt files recorded in the three stages (warmup, search, and finetuning stage), as well as .ckp files for the three stages (e.g. final_best_search.ckp, final_best_finetuning.ckp, final_best-warmup.ckp). I would like to know how to evaluate the results and compare them with the results of quantization and pruning performed separately or in sequence. Is there any parameter that can make the results more visual? thanks.

2. I only ran the file (icl_mixprec_training_cost_size.py) once, and the result was .pt and .ckp files. Is there something wrong with my process?

**Comment by bmots** [comment]
Hello @tang12138-yyhh,

1. In the log file that is created by the script you can find the regularization metrics of interest (e.g. size of the DNN, number of inference cycles, energy) and the accuracy on the benchmark. To compare the results of the joint methodology with the sequential application of pruning and MPQ you can create, for instance, a plot showing the Pareto front of the obtained architectures in the cost (size or a hardware-specific metric) vs accuracy space, as in Figure 5 of [our paper](https://arxiv.org/pdf/2407.01054). You can additionally evaluate the results  in terms of required training time, as explained in Sec. 5.3 of the paper.
2. To obtain an entire Pareto front of architectures, you have to run the script multiple times varying the regularization strength hyperparameter, that controls the trade-off between the task loss and regularization loss terms. You can do this by regulating the `strength` argument in the script.

I hope this answers your questions.

**Comment by tang12138-yyhh** [comment]
Hello, @bmots 
![1](https://github.com/user-attachments/assets/a7226f3e-a326-4539-8c56-32ec267ce4b0)
figure 1
Figure 1 shows that the alpha summary during the warmup phase looks normal.

![2](https://github.com/user-attachments/assets/d8d643d9-71fc-43e3-b2cb-829b95c3f698)
However, as shown in Figures 2, the model size remains 0 during the search and fine-tuning stages. I don't know where the problem lies.

Question: These are the parameters I set. Is it possible that the problem is here? If so, could you please correct me? thank you.
![image](https://github.com/user-attachments/assets/5557dc14-1937-4d6c-aeab-bcdb3f1a691d)

**Comment by bmots** [comment]
Hello @tang12138-yyhh,

1. Tha alpha summary shown in Figure 1 does not refer to the warmup phase, but it is the alpha initialization at the beginning of the search phase. In the warmup phase the quantization parameters do not exist yet, since the warmup is just used to obtain a good weights initialization for the seed model, without optimizing its architecture yet.
2. That may be related to a too high regularization strength. The alpha summary shown in Figure 2 shows indeed that the quantization parameters associated to the 0-bit precision (i.e. pruning) have the highest values overall. You can try to assign lower values to the regularization strength hyperparameter. The other parameters that you show do not cause the issue, they are absolutely reasonable.

**Comment by tang12138-yyhh** [comment]
Thanks, @bmots 

Thank you for your reply! But I still have a few questions for you:

1. In the log file, in each epoch, what do in_a_precision and out_a_precision mean? (In my opinion, this refers to input accuracy and output accuracy, as shown in Fig.2 in the paper)
![image](https://github.com/user-attachments/assets/652492be-6e36-4b90-80b5-ac338689bc53)
2. Precision allocation of weights and activation values. For example, the DNN network in the current layer, for example, the weight and activation candidate precision sets are {0, 2, 4, 8} and {2, 4, 8}, respectively. What method is used to determine how many bits should be selected for the current layer?

3. Assuming I understand correctly, pruning in this article should mean quantifying the weight to 0bit?

**Comment by bmots** [comment]
Hello @tang12138-yyhh,

1. Figure 2 of the paper does not show the input/output accuracy. It rather delineates the quantization approach within each layer of the DNN.  In the log, “in_a_precision” and “out_a_precision” refer to the bit-width for the input and output activations that would currently be selected according to the value of the quantization parameters at such step of the optimization process.
2. The optimization method that is used by our algorithm is detailed in Section 4.2 of the [paper](https://arxiv.org/abs/2407.01054). Let us know if you have any specific doubts about it.
3. The core idea of the algorithm is that pruning a channel can be seen as quantizing it to 0-bit precision, i.e. effectively removing its contribution from the computations.
